core/src/main/java/io/parsingdata/metal/data/Slice.java

```diff
@@ -31,24 +34,29 @@
     public final long offset;
     public final BigInteger length;
 
-    public Slice(final Source source, final long offset, final BigInteger length) {
+    private Slice(final Source source, final long offset, final BigInteger length) {
         this.source = checkNotNull(source, "source");
         this.offset = offset;
         this.length = checkNotNull(length, "length");
     }
 
+    public static Optional<Slice> createFromSource(final Source source, final long offset, final BigInteger length) {
```

Comment:
Does this factory method need argument checking?

Reply:
Is a negative offset valid? I am a bit confused, perhaps isAvailable handles those cases (I'd prefer to fail early), but you do  specifically check for the length here.